File: Data Structures & Algorithms/search-2d-matrix/submission-4.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int l=0,r=matrix.size()-1,mid,m,b=matrix[0].size()-1;
        while(l<=r){
            mid = (l+r)/2;
            if(matrix[mid][0]<=target && matrix[mid][b]>=target){
                l = 0;
                while(l<=b){
                    m = (l+b)/2;
                    if(matrix[mid][m]==target){
                        return true;
                    }else if(matrix[mid][m]<target){
                        l = m + 1;
                    }else{
                        b = m - 1;
                    }
                }
                return false;
            }else if(matrix[mid][0]>target){
                r = mid-1;
            }else{
                l = mid + 1;
            }
        }
        return false;
    }
};

File: Data Structures & Algorithms/4sum/submission-4.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        Set<List<Integer>> set = new LinkedHashSet<>();
        for(int i=0;i<nums.length;i++){
            for(int j=i+1;j<nums.length;j++){
                int l=j+1;
                int r=nums.length-1;
                while(l<r){
                    long sum=(long)nums[i]+nums[j]+nums[l]+nums[r];
                    if(sum==target){
                        set.add(Arrays.asList(nums[i],nums[j],nums[l],nums[r]));
                        l++;
                        r--;
                    }

                    if(sum<target){l++;}
                    if(sum>target){r--;}
                }
            }
        }
        return new ArrayList<>(set);
    }
}